### ISSUE TITLE
How to disable apiserver-kicker?  (And why shouldn't I?)

### ISSUE THREAD

**Issue by robotrapta** [canonical/microk8s#2790]
Frequently after I reboot my single-node cluster, a few minutes after kubernetes appears to be settled and working properly, the apiserver-kicker decides to kill the kubelet, and then the pods which were happily doing useful work get unceremoniously killed with `SandboxChanged`.

So right now, I'd just like to turn off the apiserver-kicker permanently.  What's the right way to do that?  Is it just `sudo systemctl disable snap.microk8s.daemon-apiserver-kicker.service`?  I'm also open to feedback as to why I would actually want to leave that thing on.

**Comment by ktsakalozos** [comment]
Hi @robotrapta

The behavior that most probably causes the cluster restart is a network change. When MicroK8s on a single detects a network change (eg an IP change, a new network interface etc) the apiserver-kicker will trigger a reconfiguration of the cluster that results in services restart. The use case we want to accommodate with this behavior is for example a laptop moving from network to network or a workstation gets a new IP from DHCP after a reboot. If you are certain you do not need MicroK8s to follow network changes add the `--advertise-address 0.0.0.0` option to `/var/snap/microk8s/current/args/kube-apiserver` and restart one last time MicroK8s with `microk8s stop; microk8s start`.

**Comment by robotrapta** [comment]
Thanks for the pointer.  Yeah, the log says it detected a network change - probably the VPN connecting.  But no kubernetes traffic uses that VPN, so I'd really like it to ignore that.

So if I specify the apiserver address explicitly, then the apiserver-kicker won't restart the kubelet on network changes?

**Comment by adamdunstan** [comment]
https://github.com/canonical/microk8s/issues/1943

This issue was discussed here as well.  The API kicker is an indiscriminate tool, it makes the assumption that only microk8s depends on the hosts interface, this is incorrect.  Again, I suggest that the maintainers look at the ticket above.  The kicker should only run when addresses that microk8s is directly using, not when any interface has changed.  I am about to open another ticket on this issue.  When using a VM and the other end of the veth pair is a bridge, the kicker fires continously.

**Comment by Gregro** [comment]
I'm new to Ubuntu Core and microk8s and really struggled with this over the past week, nearly to the point of bailing out of both.  I installed Ubuntu Core 22 on an Intel NUC w/ a static IPv4 via ethernet, "disabled" IPv6, and left WiFi unconfigured.  

My pods kept randomly restarting.  Even though I saw new network interfaces showing up via `networkctl`, it took me a long time to find this issue and figure out that apiserver-kicker was the problem.  I still don't know what "network changes" it was detecting...

Adding `--advertise-address 0.0.0.0` to the top of `/var/snap/microk8s/current/args/kube-apiserver` and restarting fixed it, so thanks for that tip! 

That might be a good thing to add to [Getting started with MicroK8s on Ubuntu Core](https://ubuntu.com/tutorials/getting-started-with-microk8s-on-ubuntu-core?_ga=2.72141395.1186792598.1675626007-1124420606.1675626007#4-start-microk8s-and-check-the-status) or [Microk8s Troubleshooting](https://microk8s.io/docs/troubleshooting).  If my vanilla networking setup triggered this, I'm sure others will encounter it.
